Use the GNU Gold or LLVM LLD linker if available

nilsdeppe opened the pull request:
## Proposed changes

Use the GNU Gold and LLVM LLD linkers if they're available.

These linkers are quite a bit faster than LD. On my machine I get:
ld: 11.7 s
ld.gold: 6.5 s
ld.lld: 4.9 s

when linking RunTests

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

nilsdeppe committed:
Use the GNU Gold or LLVM LLD linker if available

These linkers are quite a bit faster than LD. On my machine I get:
ld: 11.7 s
ld.gold: 6.5 s
ld.lld: 4.9 s

when linking RunTests

wthrowe reviewed (changes requested):
There needs to be a way for the user to override this choice.  I've used systems where ld.gold would successfully link binaries but those binaries wouldn't run.

nilsdeppe commented:
I've updated the commit. Given that it's a single file I just squashed since that's probably easier to read than a diff. Let me know if you think the option name should be changed, I went with what the underlying flag to the compiler is.
